### PR TITLE
Keep `preds` and `targets` as torch tensor during accuracy evaluation

### PR DESCRIPTION
**Pull request by jambayk**
## Describe your changes
During accuracy evaluation, we convert the output and target tensors to list before computing the metric value. However, this method is expensive and not efficient when the tensors are large since we end up converting them back to torch tensors anyways. 
This is felt significantly when evaluating perplexity for large language models where the `preds` tensor is a large logit tensor of size `batch_size, seqlen, vocab` (like `1 X 2048 X 32000`). This is very slow. 

In this PR, we just move the tensors to `cpu` if not already on cpu and concatenate them using `torch.cat`. This has the same effect as `list.extend`. 

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Format your code by running `pre-commit run --all-files`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.

### CODE REVIEW

File: olive/evaluator/accuracy.py [hunk 1]
```diff
@@ -62,6 +61,13 @@ def _metric_config_from_torch_metrics(cls):
     def _default_config(cls) -> Dict[str, ConfigParam]:
         return cls._metric_config_from_torch_metrics()
 
+    @staticmethod
+    def prepare_tensors(preds, target, dtypes=torch.int):
+        dtypes = dtypes if isinstance(dtypes, (list, tuple)) else [dtypes, dtypes]
```
Comment (trajepl) [inline olive/evaluator/accuracy.py:66]: how about to add assertion:
assert len(dtypes) == 2

Comment (guotuofeng) [inline olive/evaluator/accuracy.py:66]: could we just use the dtype instead of [dtypes, dtypes]?

Comment (jambayk) [reply olive/evaluator/accuracy.py:66]: done!

Comment (jambayk) [reply olive/evaluator/accuracy.py:66]: The idea was to support either a len2 tuple/list or a single type. So that the types for preds and tensor can be different if wanted. 
I am using the `[dtypes, dtypes]` part to make the logic simpler (no additional conditions) for the following lines that refer to `dtypes[0]` and `dtypes[1]`.

File: olive/evaluator/olive_evaluator.py [hunk 2]
```diff
@@ -334,8 +334,11 @@ def _inference(
                 # convert to dict of torch tensor
                 result = {name: torch.Tensor(res[i]) for i, name in enumerate(output_names)}
             outputs = post_func(result) if post_func else result
-            preds.extend(outputs.tolist())
-            targets.extend(labels.data.tolist())
+            # keep as numpy or torch arrays
+            preds.append(outputs.cpu())
```
Comment (guotuofeng) [inline olive/evaluator/olive_evaluator.py:338]: the append will make the list as a single item. That's the why the extend is used originally.

Comment (jambayk) [reply olive/evaluator/olive_evaluator.py:338]: torch.cat at line 340 and 341 does the same thing as noted in the PR description

**Review by trajepl — APPROVED** [review]
